workers/unity/Packages/io.improbable.worker.sdk/EventTracing.cs

```diff
@@ -0,0 +1,349 @@
+        {
+            var createdSpan = CEventTrace.EventTracerAddSpan(eventTracer, null, (uint) causeCount);
+        }
+
+        public void AddEvent(Event @event)
+        {
+            CEventTrace.EventTracerAddEvent(eventTracer, ParameterConversion.ConvertEvent(@event));
+        }
+
+        public bool ShouldSampleEvent(Event @event)
+        {
+            return CEventTrace.EventTracerShouldSampleEvent(eventTracer, ParameterConversion.ConvertEvent(@event)) > 0;
+        }
+    }
+
+    public unsafe class TraceEventData : IDisposable
+    {
+        internal GCHandle eventData;
+
+        private readonly GcHandlePool fieldHandles;
+
+        public TraceEventData()
+        {
+            var internalEventData = CEventTrace.EventDataCreate();
+            eventData = GCHandle.Alloc(internalEventData, GCHandleType.Pinned);
+            fieldHandles = new GcHandlePool();
+        }
+
+        public void Dispose()
+        {
+            if (eventData.IsAllocated)
+            {
+                eventData.Free();
+            }
+
+            fieldHandles.Dispose();
+        }
+
+        public void AddCollectionOfFields(Dictionary<string, string> fields)
+        {
+            foreach (var entry in fields)
+            {
+                AddField(entry.Key, entry.Value);
+            }
+        }
+
+        public void AddField(string key, string value)
+        {
+            var pinnedKey = fieldHandles.Pin(ApiInterop.ToUtf8Cstr(key));
+            var pinnedValue = fieldHandles.Pin(ApiInterop.ToUtf8Cstr(value));
+
+            CEventTrace.EventDataAddStringFields(eventData.AddrOfPinnedObject(), 1,
+                (byte**) pinnedKey.ToPointer(), (byte**) pinnedValue.ToPointer());
+        }
+
+        public Dictionary<string, string> GetAllFields()
+        {
+            var numberOfFields = CEventTrace.EventDataGetFieldCount(eventData.AddrOfPinnedObject());
+            var nativeKeys = new byte*[numberOfFields];
+            var nativeValues = new byte*[numberOfFields];
+
+            fixed (byte** keys = nativeKeys)
+            fixed (byte** values = nativeValues)
+            {
+                CEventTrace.EventDataGetStringFields(eventData.AddrOfPinnedObject(), keys, values);
+            }
+
+            var fields = new Dictionary<string, string>();
+            for (var i = 0; i < numberOfFields; i++)
+            {
+                fields.Add(ApiInterop.FromUtf8Cstr(nativeKeys[i]), ApiInterop.FromUtf8Cstr(nativeValues[i]));
+            }
+
+            return fields;
+        }
+
+        public string GetFieldValue(string key)
+        {
+            byte* value;
+            fixed (byte* nativeKey = ApiInterop.ToUtf8Cstr(key))
+            {
+                value = CEventTrace.EventDataGetFieldValue(eventData.AddrOfPinnedObject(), nativeKey);
+            }
+
+            return ApiInterop.FromUtf8Cstr(value);
+        }
+    }
+
+    public unsafe class TraceItem
```

Comment:
It seems that most of these methods belong on the `Item` struct that then take in `IOStorage/IOStream` as required.

For example:

```csharp
public struct Item 
{
	// Field declarations
	
	public void AddToStorage(IOStorage storage);
	public void SerializeToStream(IOStream stream);
	public long GetSerializedSize(); // Does this need to be public? 

	public static Item DeserializeFromStream(IOStream stream);
}
```

This feels more natural than having a secondary class which takes in both a storage and stream and maybe will use them depending on what you do. Unless I'm misunderstanding what this class is for?

Reply:
I've move the method previously in `TraceItem` into the `Item` struct. Because of the way the Item conversion from user-facing to internal works, I needed to add the IOStorage parameter to some methods where it doesn't seem like you would need it. i.e. `SerializeItemToStream`